grr/server/grr_response_server/databases/spanner_utils.py

```diff
@@ -0,0 +1,477 @@
+
+  def Transact(
+      self,
+      func: Callable[["Transaction"], _T],
+      txn_tag: Optional[str] = None,
+  ) -> List[Any]:
+
+    """Execute the given callback function in a Spanner transaction.
+
+    Args:
+      func: A transaction function to execute.
+      txn_tag: Transaction tag to apply.
+
+    Returns:
+      The result of the transaction function executed.
+    """
+    return self._pyspanner.run_in_transaction(func, transaction_tag=txn_tag)
+
+  def Mutate(
+      self, func: Callable[["Mutation"], None], txn_tag: Optional[str] = None
+  ) -> None:
+    """Execute the given callback function in a Spanner mutation.
+
+    Args:
+      func: A mutation function to execute.
+      txn_tag: Optional[str] = None,
+    """
+
+    self.Transact(func, txn_tag=txn_tag)
+
+  def Query(self, query: str, txn_tag: Optional[str] = None) -> Cursor:
+    """Queries Spanner database using the given query string.
+
+    Args:
+      query: An SQL string.
+      txn_tag: Spanner transaction tag.
+
+    Returns:
+      A cursor over the query results.
+    """
+    with self._pyspanner.snapshot() as snapshot:
+        results = snapshot.execute_sql(query, request_options={"request_tag": txn_tag})
+
+    return results
+
+  def QuerySingle(self, query: str, txn_tag: Optional[str] = None) -> Row:
+    """Queries PySpanner for a single row using the given query string.
+
+    Args:
+      query: An SQL string.
+      txn_tag: Spanner transaction tag.
+
+    Returns:
+      A single row matching the query.
+
+    Raises:
+      NotFound: If the query did not return any results.
+      ValueError: If the query yielded more than one result.
+    """
+    return self.Query(query, txn_tag=txn_tag).one()
+
+  def ParamQuery(
+      self, query: str, params: Mapping[str, Any],
+      param_type: Optional[dict] = None, txn_tag: Optional[str] = None
+  ) -> Cursor:
+    """Queries PySpanner database using the given query string with params.
+
+    The query string should specify parameters with the standard Python format
+    placeholder syntax [1]. Note that parameters inside string literals in the
+    query itself have to be escaped.
+
+    Also, the query literal is not allowed to contain any '@p{idx}' strings
+    inside as that would lead to an incorrect behaviour when evaluating the
+    query. To prevent mistakes the function will raise an exception in such
+    cases.
+
+    [1]: https://docs.python.org/3/library/stdtypes.html#str.format
+
+    Args:
+      query: An SQL string with parameter placeholders.
+      params: A dictionary mapping parameter name to a value.
+      txn_tag: Spanner transaction tag.
+
+    Returns:
+      A cursor over the query results.
+
+    Raises:
+      ValueError: If the query contains disallowed sequences.
+      KeyError: If some parameter is not specified.
+    """
+     if not param_type:
+      param_type = {}
+    names, values = collection.Unzip(params.items())
```

Comment:
```suggestion
    names, _ = collection.Unzip(params.items())
```

`values` isn't being used.